server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -0,0 +1,362 @@
+     */
+    protected final boolean countCanUseMetadata(FiltersAggregator.Counter counter, Bits live) {
+        if (live != null) {
+            /*
+             * We can only use metadata if all of the documents in the reader
+             * are visible. This is done by returning a null `live` bits. The
+             * name `live` is traditional because most of the time a non-null
+             * `live` bits means that there are deleted documents. But `live`
+             * might also be non-null if document level security is enabled.
+             */
+            return false;
+        }
+        /*
+         * We can only use metadata if we're not using the special docCount
+         * field. Otherwise we wouldn't know how many documents each lucene
+         * document represents.
+         */
+        return counter.docCount.alwaysOne();
+    }
+
+    /**
+     * Make a filter that matches this filter and the provided query.
+     * <p>
+     * Note: This method rewrites the query against the {@link IndexSearcher}.
+     */
+    QueryToFilterAdapter<?> union(Query extraQuery) throws IOException {
+        extraQuery = searcher().rewrite(extraQuery);
+        if (extraQuery instanceof MatchAllDocsQuery) {
+            return this;
+        }
+        Query unwrappedQuery = unwrap(query);
+        Query unwrappedExtraQuery = unwrap(extraQuery);
+        if (unwrappedQuery instanceof PointRangeQuery && unwrappedExtraQuery instanceof PointRangeQuery) {
+            Query merged = MergedPointRangeQuery.merge((PointRangeQuery) unwrappedQuery, (PointRangeQuery) unwrappedExtraQuery);
+            if (merged != null) {
+                // Should we rewrap here?
+                return new QueryToFilterAdapter<>(searcher(), key(), merged);
+            }
+        }
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        builder.add(query, BooleanClause.Occur.MUST);
+        builder.add(extraQuery, BooleanClause.Occur.MUST);
+        return new QueryToFilterAdapter<>(searcher(), key(), builder.build());
+    }
+
+    private static Query unwrap(Query query) {
+        while (true) {
+            if (query instanceof ConstantScoreQuery) {
+                query = ((ConstantScoreQuery) query).getQuery();
+                continue;
+            }
+            if (query instanceof IndexSortSortedNumericDocValuesRangeQuery) {
+                query = ((IndexSortSortedNumericDocValuesRangeQuery) query).getFallbackQuery();
+                continue;
+            }
+            if (query instanceof IndexOrDocValuesQuery) {
+                query = ((IndexOrDocValuesQuery) query).getIndexQuery();
+                continue;
+            }
+            return query;
+        }
+    }
+
+    /**
+     * Build a predicate that the "compatible" implementation of the
+     * {@link FiltersAggregator} will use to figure out if the filter matches.
+     * <p>
+     * Consumers of this method will always call it with non-negative,
```

Comment:
```suggestion
     * Consumers of the returned predicate will always call it with non-negative,
```
I found the pronoun ambiguous in this context.